samza-shell/src/main/bash/run-class.sh

```diff
@@ -45,19 +45,68 @@ DEFAULT_LOG4J2_FILE=$base_dir/lib/log4j2.xml
 BASE_LIB_DIR="$base_dir/lib"
 # JOB_LIB_DIR will be set for yarn container in ContainerUtil.java
 # for others we set it to home_dir/lib
-JOB_LIB_DIR="${JOB_LIB_DIR:-$base_dir/lib}"
+JOB_LIB_DIR="${JOB_LIB_DIR:-$home_dir/lib}"
 
 export JOB_LIB_DIR=$JOB_LIB_DIR
 
-echo JOB_LIB_DIR=$JOB_LIB_DIR
-echo BASE_LIB_DIR=$BASE_LIB_DIR
+# Usage: relative_path_from_to $source $target
+# Purpose: Given two absolute paths return relative path from $source to $destination
+function relative_path_from_to() {
+ source=${1%/} # This is $home_dir for this case
+ target=${2%/} # This is $BASE_LIB_DIR where JARS are
+
+ common_part=$source # start with common_part as source and iterate until common_part becomes $target
+ result=""
+
+ # This calculates the longest common prefix
+ while [[ "${target#$common_part}" == "${target}" ]]; do
```

Comment:
Though this idea sounds interesting, can we just cd into the  directory(BASE_LIB_DIR) that has all the jars and run the java launch command with absolute jar path. 

Something along the lines of the following:
```bash 
sh -c 'cd BASE_LIB_DIR &&  exec $JAVA $JAVA_OPTS -cp $CLASSPATH "$@"
```
I have assumed that BASE_LIB_DIR has all the relevant jars. I think this will preserve the current working dir in the script and simplify the code in `run-class.sh`. 

Will that work in this case?

Reply:
This will trouble us when we want to read jar/wars out of JOB_LIB_DIR and BASE_LIB_DIR

Reply:
Even in your current solution, you're using relative_path for one lib directory and absolute path for the other lib directory. You can follow the same approach with the above idea as well. I think it'll be good for us in the long run to go with a simpler solution rather than computing longest common prefix and stripping stuff. 

If you disagree, then please share your rationale for doing so?

Reply:
- Are you sure `sh -c 'cd BASE_LIB_DIR &&  exec $JAVA $JAVA_OPTS -cp $CLASSPATH'`
works? Please confirm on this
- Also please throw more light on the issues with the current solution? Why do you think it will be a problem?
- One more thing to mention: Look at the shebang, might run into syntax errors on using sh 

Please be considerate that we have spent time and effort thoroughly testing/reviewing the current fix in various deployment modes.

Reply:
>> Are you sure sh -c 'cd BASE_LIB_DIR && exec $JAVA $JAVA_OPTS -cp $CLASSPATH'
works? Please confirm on this

It's the contract of [GNU shell](https://linux.die.net/man/1/sh).


>> Also please throw more light on the issues with the current solution? Why do you think it will be a problem?
1. Existing solution is hard to understand of me now. In the middle of the night, admist a GCN, I think it wouldn't be a great to decipher the logic in it. 
2. Any other developer who wants to alter it in the future would have a tough time with it.


>> One more thing to mention: Look at the shebang, might run into syntax errors on using sh

There are multiple ways to solve this problem. Take a look [here](https://stackoverflow.com/questions/10382141/temporarily-change-current-working-directory-in-bash-to-run-a-command). 
I am not suggesting that you should switch to the approach I mentioned. At-least we should try to simplify things here by trying alternate options. I just suggested an alternative. 


>> Please be considerate that we have spent time and effort thoroughly testing/reviewing the current fix in various deployment modes.

IMHO, this is not a strong enough rationale.  I am not going to stand in the way if everyone else thinks this is fine.

Reply:
Tested with `cd BASE_LIB_DIR && exec $JAVA $JAVA_OPTS -cp $CLASSPATH` and found this breaks control-dev & control-yarn, since they build bash arguments expecting this execution of JAVA from a specific path. We use relative path in those scripts for certs, other config files.